### ISSUE TITLE
wish: add key to kill-ring-save the inferred type

### ISSUE THREAD

**Issue by nahuel** [emacs-lsp/lsp-mode#1708]
Many times I want to type an inferred type, it will be nice to have a shorcut to copy the type below the cursor. Typescript example:

```ts
type SomeType = 'q' | 'w'
const b = () => 'q' as SomeType

function a(arg /* here I want to type " : " and then paste "SomeType" */ ) {
  const x = b() 
     // ^ Cursor here
}
``` 

Here cursor is over `x`, so the inferred type is shown in the minibuffer as `const x: SomeType`. Now it will be nice to press a (new) shorcut that copies the `x` type name to the clipboard, so then you can move the cursor after `arg`, type ` : ` and then <kbd>C-Y</kbd> to paste `SomeType`, to declare the type of `arg` as the same of `x`.

**Comment by yyoncho** [comment]
> so the inferred type is shown in the minibuffer as const x: SomeType

The issue is that this information is not structured and might not be the type itself. I think that the only reliable server-side method for that matter is documentSymbols call(corresponds to `imenu`). We could use it to copy the name of the wrapping function, copy the name of the wrapping class, etc.

**Comment by nahuel** [comment]
Can the mechanism behind `lsp-rust-analyzer-inlay-hints-mode` be reused for getting the expression type in a way generic for all LSP backends? (note, I didn't read the LSP spec)

**Comment by nahuel** [comment]
Note, I filled a similar wish for vscode: https://github.com/microsoft/vscode/issues/98267

**Comment by yyoncho** [comment]
> Can the mechanism behind `lsp-rust-analyzer-inlay-hints-mode` be reused for getting the expression type in a way generic for all LSP backends?

This is RA specific. We could provide a mechanism to copy the inlay-hint on the line, this will be easy, but it will work only for rust analyzer.

**Comment by nahuel** [comment]
Another option maybe is to parse what the minibuffer shows, at least for `const` / `let` declarations, and extract the type from here. FYI just found a proposal from the Rust people to add inlay hints to the LSP spec: https://github.com/microsoft/language-server-protocol/issues/956 . But maybe the LSP spec needs a more general "get the type at position" functionality than that?

**Comment by yyoncho** [comment]
> But maybe the LSP spec needs a more general "get the type at position" functionality than that?

Yep, in order to work consistently. 

> Another option maybe is to parse what the minibuffer shows, at least for const / let declarations,

I tend to avoid such ad-hoc solutions because they cause trouble. After all, lsp protocol is to avoid client side parsing.

**Comment by nahuel** [comment]
ok, I understand. Note, others are finding this same LSP limitation and for a more general usage, as getting the type of a selected expression instead of only of the symbol at cursor: 
https://github.com/rust-analyzer/rust-analyzer/issues/389
https://github.com/digital-asset/ghcide/issues/70

**Comment by yyoncho** [comment]
yep, unfortunately, the spec is designed for point-and-click usage.

**Comment by yyoncho** [comment]
closing based on the previous comment. Once the spec adds support for it we can implement the client-side code.
